### PR TITLE
Make all Element with Template pickable natively

### PR DESCRIPTION
**Pull request by BastienGauthier**
Add the template_str attribute to be able to recover the Element after unpickling thanks to `__setstate__` (needed to pickle Map in folium https://github.com/python-visualization/folium/pull/1812).

### CODE REVIEW

File: branca/element.py [hunk 1]
```diff
@@ -57,11 +57,13 @@ def __init__(self, template=None, template_name=None):
         self._env = ENV
         self._children = OrderedDict()
         self._parent = None
+        self.template = template
+        self.template_name = template_name
```
Comment (Conengmo) [inline branca/element.py:61]: I'm thinking whether it's better to make these protected attributes. To signal to users they are not meant for them. And to possibly avoid confusion between `_template` and `template` here.

How about  perhaps`self._template_str` and `self._template_name`?

File: branca/element.py [hunk 2]
```diff
@@ -57,11 +57,13 @@ def __init__(self, template=None, template_name=None):
...
 
-        if template is not None:
-            self._template = Template(template)
-        elif template_name is not None:
-            self._template = ENV.get_template(template_name)
+        if self.template is not None:
+            self._template = Template(self.template)
+        elif self.template_name is not None:
+            self._template = ENV.get_template(self.template_name)
```
Comment (Conengmo) [inline branca/element.py:66]: Could you revert the changes to these lines? They are not really doing any meaningful change, so in that case I'd prefer to keep the diff smaller. Smaller diffs means less chances for mistakes.

File: branca/element.py [hunk 3]
```diff
@@ -71,11 +73,20 @@ def __getstate__(self):
...
         state: dict = self.__dict__.copy()
         state["_env"] = None
+        state["_template"] = None
         return state
 
     def __setstate__(self, state: dict):
         """Re-add ._env attribute when unpickling"""
         state["_env"] = ENV
+
+        state["_template"] = self._template
```
Comment (Conengmo) [inline branca/element.py:83]: This is a bit of a small optimization. But I'm thinking how to make it clearer what is going on here. How about instead we remove `_template` if it exists. That way we don't have to overwrite it here with `self._template`.

So on line 76: `state.pop('_template', None)`
And remove line 83.

If I'm not mistaken that should be equivalent to the current change, but with one less action.

**Review by Conengmo — CHANGES_REQUESTED** [review]
Great how you got this change down to just the stuff in Element!

File: branca/element.py (hunk 1 above)
Comment (BastienGauthier) [reply branca/element.py]: Done !

File: branca/element.py (hunk 2 above)
Comment (BastienGauthier) [reply branca/element.py]: Done !

File: branca/element.py (hunk 3 above)
Comment (BastienGauthier) [reply branca/element.py]: Took me a while to understand, but ok, resolved now !

**Comment by BastienGauthier** [comment]
Ok here too !

**Review by Conengmo — APPROVED** [review]
Really great how this turned out! Thanks for the journey @BastienGauthier.
